packages/vite/src/node/utils.ts

```diff
@@ -1438,3 +1438,34 @@ export function partialEncodeURIPath(uri: string): string {
   const postfix = filePath !== uri ? uri.slice(filePath.length) : ''
   return filePath.replaceAll('%', '%25') + postfix
 }
+
+export const closeServerAndExit = async (
+  server: ViteDevServer | PreviewServer | PreviewServer['httpServer'],
+): Promise<void> => {
+  try {
+    await server.close()
+  } finally {
+    process.exit()
+  }
+}
+
+export const createCloseServerAndExitFn =
+  (server: ViteDevServer | PreviewServer): (() => Promise<void>) =>
+  () =>
+    closeServerAndExit(server)
```

Comment:
I think we can avoid these two abstractions and directly inline the code. It is just a little bit more verbose but we don't need to check what is going on inside.

Reply:
I agree about `createCloseServerAndExitFn()`, we can inline that and avoid the mental load of the closure, but could we just keep `closeServerAndExit()` ? It is used in 4 places (shortcuts + dev and preview server), I find it slightly better to avoid having the same piece of code 4 times.

Reply:
I the overloading a bit complex. The explicit try finally block very clear to me, and the repetition doesn't bothers the reader.

Reply:
Addressed in 9e32ee2. `closeServerAndExit()` in index and preview server is needed so the same reference can be passed in setup and teardown functions.

Reply:
[This run](https://github.com/vitejs/vite/actions/runs/9404508531/job/25903574897?pr=17333) looks like a flake, it's passing locally.